src/olympia/api/tests/test_permissions.py

```diff
@@ -54,7 +75,193 @@ class TestGroupPermission(TestCase):
+        assert AnyOf(AllowNone, AllowAny)().has_object_permission(
+            request, myview, None)
+        assert AnyOf(AllowAny, AllowNone)().has_object_permission(
+            request, myview, None)
+
+    def test_has_object_permission_fail(self):
+        request = RequestFactory().get('/')
+        assert not AnyOf(AllowNone, AllowNone)().has_object_permission(
+            request, myview, None)
+
+    def test_has_object_permission_partial_fail(self):
+        """Test that AnyOf.has_object_permission() does not allow access when
+        a permission class returns False for has_permission() without having
+        a has_object_permission() implementation."""
+
+        class NoObjectPerm(BasePermission):
+            # This class will not grant access because we do check
+            # has_permission() on top of just has_object_permission().
+            def has_permission(self, request, view):
+                return False
+
+        class NoPerm(BasePermission):
+            # This class will not grant access either when checking
+            # has_object_permission() since it directly returns False.
+            def has_object_permission(self, request, view, obj):
+                return False
+
+        request = RequestFactory().get('/')
+        assert not AnyOf(NoObjectPerm, NoPerm)().has_object_permission(
+            request, myview, None)
+
+
+class TestAllowAddonAuthor(TestCase):
+    fixtures = ['base/users', 'base/addon_3615']
+
+    def setUp(self):
+        self.addon = Addon.objects.get(pk=3615)
+        self.permission = AllowAddonAuthor()
+        self.owner = self.addon.authors.all()[0]
+        self.request = RequestFactory().get('/')
+        self.request.user = AnonymousUser()
+
+    def test_has_permission_anonymous(self):
+        assert not self.permission.has_permission(self.request, myview)
+
+    def test_has_permission_any_authenticated_user(self):
+        self.request.user = UserProfile.objects.get(pk=999)
+        assert self.request.user not in self.addon.authors.all()
+        assert self.permission.has_permission(self.request, myview)
+
+    def test_has_object_permission_user(self):
+        self.request.user = self.owner
+        assert self.permission.has_object_permission(
+            self.request, myview, self.addon)
+
+    def test_has_object_permission_different_user(self):
+        self.request.user = UserProfile.objects.get(pk=999)
+        assert self.request.user not in self.addon.authors.all()
+        assert not self.permission.has_object_permission(
+            self.request, myview, self.addon)
+
+    def test_has_object_permission_anonymous(self):
+        assert not self.permission.has_object_permission(
+            self.request, myview, self.addon)
+
+
+class TestAllowReviewer(TestCase):
+    fixtures = ['base/users']
+
+    # Note: be careful when testing, under the hood we're using a method that
+    # relies on UserProfile.groups_list, which is cached on the UserProfile
```

Comment:
It's cached on the instance inside `__dict__` so it's not necessary. I added the warning in case people add tests that access the group list a few times in the same test.

For instance, the following test would fail on the second assert:
```
request.user = UserProfile.objects.get(pk=999)
assert not self.permission.has_permission(request, myview)
GroupUser.objects.create(user=request.user, group=reviewers)
assert self.permission.has_permission(request, myview)
```